yulopti: Add support for Yul Objects

dflupu opened the pull request:
This fixes [#12856](https://github.com/ethereum/solidity/issues/12856).

~I've made a small change to ObjectParser as part of this to set the `.subId` fields as they didn't seem to be set anywhere.~

This is my first contribution to this project so be aware of that when reviewing the changes.

dflupu committed:
yulopti: Add support for Yul Objects

chriseth commented on libyul/ObjectParser.cpp:
```diff
@@ -100,7 +100,10 @@ shared_ptr<Object> ObjectParser::parseObject(Object* _containingObject)
 	}
 	if (_containingObject)
+	{
+		ret->subId = _containingObject->subObjects.size();
```
Currently, this is only set in the EVM object compiler. I think it is a reference to something from Assembly.cpp, so I don't think we should set it to an arbitrary value here.

dflupu replied:
Got it, I've removed this line and the use of `pathToSubObject`. The value I had set was required for `pathToSubObject` to work.

chriseth commented on test/tools/yulopti.cpp:
```diff
@@ -265,10 +369,11 @@ int main(int argc, char** argv)
 	interactively read from stdin.
 	In non-interactive mode a list of steps has to be provided.
 	If <file> is -, yul code is read from stdin and run non-interactively.
+	If <object> is provided then only the object matching the given path will be read.
```
Can you explain what `<object>` is?

dflupu replied:
I've changed this to `An <object> flag may be provided, specifying a dotted path to an object in the input.`. Hope that's better.

chriseth commented on test/tools/yulopti.cpp:
```diff
@@ -87,32 +88,44 @@ class YulOpti
+		if (_path.empty())
+			return _rootObject;
+
+		auto pathToSubObject = _rootObject->pathToSubObject(YulString(_path));
```
`pathToSubObject` was written to be used during code generation. Because of that, it directly returns code-generation internal IDs. What we would need here is an actual Object. I think it would be better to use a separate path resolution function - maybe there are ways to share code.

dflupu replied:
Removed the use of `pathToSubObject` and simply did the search in `YulOpti::getSubObject`.

chriseth commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +184,118 @@ class YulOpti
 
+	void objectApply(function<void(Object&)> _fn)
+	{
+		vector<shared_ptr<Object>> stack;
```
I think it's totally fine to use the C++ machine stack.

dflupu replied:
I assumed this means making the function recursive. If so, done.

chriseth commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +184,118 @@ class YulOpti
+
+	void analyze(ErrorReporter& errorReporter)
+	{
+		objectApply([&](Object& object) -> void {
```
Do you need this explicit `-> void`?

dflupu replied:
Fixed

chriseth commented:
This looks good in general, but I think we are getting some complications due to the feature that we can select a single sub-object to run on. Is this actually needed? Also I didn't take a look at the output yet, but could you check that if the object is only a single object with only code, that it prints the code without the object?

dflupu commented:
> This looks good in general, but I think we are getting some complications due to the feature that we can select a single sub-object to run on. Is this actually needed? Also I didn't take a look at the output yet, but could you check that if the object is only a single object with only code, that it prints the code without the object?

Thanks for the review! I've addressed your comments and made sure that the output for a code block is still a code block. I can't comment on the utility of the `<object>` flag myself. The author of the issue might be able to. At this point the implementation for that feature is a pretty neat chunk of code that doesn't call anything weird.

leonardoalt commented:
This is still waiting for reviews. Any team member that sees this message is welcome to review it.

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -87,32 +88,68 @@ class YulOpti
+		const auto subObjectPath = _qualifiedPath.substr(_object->name.str().length() + 1);
+		const auto subObjectName = subObjectPath.substr(0, subObjectPath.find_first_of('.'));
+
+		auto subObjectIt = find_if(
```
Maybe we could use `ranges::find_if` instead of `std::find_if`.

dflupu replied:
Done

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -87,32 +88,68 @@ class YulOpti
+		auto subObjectIt = find_if(
+			_object->subObjects.begin(),
+			_object->subObjects.end(),
+			[subObjectName](auto const& subObject) { return subObject->name.str() == subObjectName; }
```
```suggestion
			[subObjectName](auto const& _subObject) { return _subObject->name.str() == subObjectName; }
```
Our code style is to use underscore for parameter variable names.

dflupu replied:
Done

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -87,32 +88,68 @@ class YulOpti
 	{
 		ErrorList errors;
 		ErrorReporter errorReporter(errors);
 		CharStream _charStream(_input, "");
```
I am aware that this was not introduced by you, but we could seize the opportunity to conform the var name to our code style, i.e, remove the underscore prefix.
@ekpyron, I saw this was introduced in commit 95c973d... maybe it was overlooked at the time, right?

dflupu replied:
Done

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +208,118 @@ class YulOpti
 		}
 	}
 
+	void objectApply(shared_ptr<Object> _object, function<void(Object&)> _fn)
```
Could we come up with a more descriptive name for this function? Apply what to object?

dflupu replied:
Renamed to `objectApplyFn`. Not sure if that's an improvement or not. Couldn't come up with anything more appropriate since it's a pretty generic function.

matheusaaguiar replied:
I think `objectApplyFunction` would be better.

dflupu replied:
Renamed to `objectApplyFunction`

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +208,118 @@ class YulOpti
+		_fn(*_object);
+	}
+
+	void analyze(ErrorReporter& errorReporter)
```
I think that the name of this function can be more explicit too. It is not clear what it is analyzing. Is is it the errors? Or the object source code?
Also, coding style.
```suggestion
	void analyze(ErrorReporter& _errorReporter)
```

dflupu replied:
Renamed to `runCodeAnalyzer`

dharmendrarawal replied:
Renamed to runCodeAnalyzer

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +208,118 @@ class YulOpti
+
+	void analyze(ErrorReporter& errorReporter)
+	{
+		objectApply(m_object, [&](Object& object) {
```
Maybe it is just me, but I think that the definition of the lambda expression here makes it a little bit hard to read, since it is fairly long. If you think it is a good idea, we could assign it to a local var before and then pass it to `objectApply` so it is more readable.
Other option, since there are breaking lines, could be to break at opening parenthesis of `objectApply`:
```suggestion
		objectApply(
			m_object,
			[&](Object& _object)
			{
```
And do we need `[&]` ? Nothing is being captured from outer scope ...
Also, coding style (underscore for param var name).

dflupu replied:
Went with the line split since the local var would've required some kind of name that'd be a duplicate of something.

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +208,118 @@ class YulOpti
-		*m_ast = std::get<yul::Block>(Disambiguator(m_dialect, *m_analysisInfo)(*m_ast));
-		m_analysisInfo.reset();
-		m_nameDispenser.reset(*m_ast);
+		objectApply(m_object, [&](Object& object) {
```
Same as before regarding breaking lines, capture and coding style. 
Please make sure to also check the next set of functions that have this lambda as a parameter of `objectApply`.

dflupu replied:
Fixed

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -214,47 +341,50 @@ class YulOpti
 						disambiguated = false;
 						break;
 					case ';':
 					{
```
I guess braces can be removed here.

dflupu replied:
Fixed

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -214,47 +341,50 @@ class YulOpti
-							std::string_view(&option, 1),
-							*m_ast
-						);
+						runSequence(std::string_view(&option, 1));
```
```suggestion
						runSequence(string_view(&option, 1));
```
We are already using namespace `std`.

dflupu replied:
Fixed

matheusaaguiar reviewed (commented):
Some suggestions and coding style fixes.

nikola-matic commented:
@dflupu heya, do you plan on continuing work on this? We're actively working on reducing the number of open PRs, and I'd like to know where we stand with this, so we can decide whether to have someone from the team to take it over, or to close it.

dflupu commented:
> @dflupu heya, do you plan on continuing work on this? We're actively working on reducing the number of open PRs, and I'd like to know where we stand with this, so we can decide whether to have someone from the team to take it over, or to close it.

I'll address the comments by the end of the week.

dflupu commented:
I think I've covered everything. Feel free to review again.

nikola-matic commented:
Heya @dflupu, thanks for picking this up again :)
A couple of pointers for future reference - when someone leaves a comment, and the fix is trivial and doesn't require a discussion (e.g. rename of a variable name), please don't leave a comment in the thread (i.e. done, fixed, etc.) as it this will send a ton of notifications and is generally annoying; in such cases, feel free to just click on resolve conversation.
Second - I've not taken a detailed look, but you've got a couple of strange failures, which will most likely be resolved if you rebase against the fresh upstream develop (e.g. the `b_archlinux` is failing due to a Z3 dependency version which was bumped some time ago).

dflupu commented:
@nikola-matic Okay! I've rebased it.

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +89,72 @@ class YulOpti
+
+		auto subObjectIt = ranges::find_if(
+			_object->subObjects,
+			[subObjectName](auto const& _subObject) { return _subObject->name.str() == subObjectName; }
```
```suggestion
			[&subObjectName](auto const& _subObject) { return _subObject->name.str() == subObjectName; }
```

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +89,72 @@ class YulOpti
+			"Assembly object not found."
+		);
+
+		const auto subObjectPath = _qualifiedPath.substr(_object->name.str().length() + 1);
+		const auto subObjectName = subObjectPath.substr(0, subObjectPath.find_first_of('.'));
```
```suggestion
		const string subObjectPath = _qualifiedPath.substr(_object->name.str().length() + 1);
		const string subObjectName = subObjectPath.substr(0, subObjectPath.find_first_of('.'));
```
I don't see any benefit of using `auto` here.

nikola-matic replied:
Also, `const` qualifier has to come after the type as per our coding style, so in this case, `string const` instead of `const string`.

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +89,72 @@ class YulOpti
+			if (!m_inputWasCodeBlock && scanner->currentToken() == Token::LBrace)
+				m_inputWasCodeBlock = true;
+
+			auto content = parser.parse(scanner, false);
+
+			if (content != nullptr)
+				m_object = getSubObject(content, _objectPath);
+
+			if (!m_object || !errorReporter.errors().empty())
```
```suggestion
			shared_ptr<Object> object = parser.parse(scanner, false);
			
			if (!object || !errorReporter.errors().empty())
```
I think it is better to check this way. If there is a parsing error, either with a `nullptr` being returned or errors present, it should be detected before continuing. 
Since `getSubObject` fails the `YulAssert` at line `117` when `nullptr` is found, it should be fine to have just the following assignment after line `151`:
``` 
m_object = getSubObject(object, _objectPath);
``` 
We could even put an assertion to guarantee `m_object` has value, but I guess it would be redundant with the one already present inside the function.

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +208,138 @@ class YulOpti
 	}
 
-	void disambiguate()
+	void objectApplyFunction(shared_ptr<Object> _object, function<void(Object&)> _fn)
```
```suggestion
	void applyFunctionToObject(shared_ptr<Object> _object, function<void(Object&)> _function)
```
I think we should definitely expand the `fn` abbreviation, and maybe the method name looks better this way?
I guess we could also consider being even more specific and using `applyFunctionToObjectAndSubObjects`... but maybe then we cross the line of readability/simplicity.

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -214,47 +361,48 @@ class YulOpti
 			}
 			cout << "----------------------" << endl;
-			cout << _source << endl;
+			printObject();
```
@cameel I think this is more a design question about the behavior of the tool. Now that it can read and work with nested objects, should their code also be printed?

cameel replied:
Yes. I haven't reviewed the changes here but I'm assuming that now the tool can work both with a Yul block (as it used to) and with a full object. In that case, with an object, the optimizer is also processing subobjects and I'd expect to see them in the output. Stripping them may give you an object that does not even compile (e.g. when you have `datacopy()` that refers to a subobject). If the user does not want to see subobjects, he can pass in a block or remove them from the input.

We may also want to add an option to ignore nested objects but I wouldn't add it until it turns out that this is something that's actually useful in practice.

matheusaaguiar replied:
@dflupu, please consider this.

dflupu replied:
This was how I had intended it to work, but it often didn't because I didn't properly initialize `m_inputWasCodeBlock` to false. Please try running the tool again and let me know if it is printing nested objects as expected now.

matheusaaguiar replied:
Ok, I confirmed that it now prints subobjects as well.

matheusaaguiar reviewed (dismissed):
I think it is good, just need to squash commits and rebase.

nikola-matic commented:
> I think it is good, just need to squash commits and rebase.

I've squashed the two fixup commits and rebase against develop. I'll give another review as well, and @matheusaaguiar, in the meantime, you can slap another approve if you're still OK with this, since the force push dismissed it.

Also, as far as squashing everything into one commit - I guess we can just use the `squash and merge` option.

nikola-matic commented on test/tools/yulopti.cpp:
```diff
@@ -265,10 +411,11 @@ int main(int argc, char** argv)
 			R"(yulopti, yul optimizer exploration tool.
 	Usage: yulopti [Options] <file>
-	Reads <file> as yul code and applies optimizer steps to it,
+	Reads <file> containing a yul object and applies optimizer steps to it,
```
Why this change? What happens if no object is present in the file?

matheusaaguiar replied:
The PR enables that objects (and nested objects) can be read by the tool. But you can have just a raw yul block too.
If there is no object nor yul block present in the file, then it terminates with an exception due to parsing error.
@dflupu , I think that it should say that a raw yul block is also accepted.

nikola-matic commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +89,70 @@ class YulOpti
 	}
 
-	void parse(string const& _input)
+	shared_ptr<Object> getSubObject(shared_ptr<Object> const& _object, string const& _qualifiedPath)
```
From what I understand, the purpose of this function is to be given a top level object, and then based on the `_qualifiedPath` to find the 'sub object' recursively; however, is it possible for the top level input object to be the the one that we've specified via the `_qualifiedPath` parameter? If yes, then this function should be called `getObject`. 

In any case, it would be good to document this function to mention that it searches for the sub object recursively based on the `_qualifiedPath` parameter.

cameel replied:
Yeah, it should generally be possible to get the top-level object too.

nikola-matic commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +89,70 @@ class YulOpti
+		if (_qualifiedPath.empty() || _qualifiedPath == _object->name.str())
+			return _object;
+
+		yulAssert(
```
I don't think the usage of `yulAssert` or any assert is appropriate here. `_qualifiedPath` is user provided and the user may make a typo, which is a perfectly valid scenario. Asserts should not be used for error handling. You could instead throw an exception here (although I'd argue that the user providing and incorrect object path is not exactly exceptional either), or even better, return a `nullptr` indicating that the object was not found, and then handle appropriately elsewhere. Same thing goes for the assert on line 110. The one on line 117 seems OK though.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +89,70 @@ class YulOpti
 	}
 
-	void parse(string const& _input)
+	shared_ptr<Object> getSubObject(shared_ptr<Object> const& _object, string const& _qualifiedPath)
```
We could add this helper to `Object`. We already have `Object::pathToSubObject()`, which overlaps some of the logic here. It could be reused if you replace the assertions with proper errors.

dflupu replied:
Added to `Object` as a static function.

Not sure what kind of errors I'm supposed to throw. `Object::pathToSubObject()` uses only asserts. For now I've made it return null if the path isn't found as suggested in another comment.

cameel replied:
It uses asserts because the caller is expected to check for these things (they're basically preconditions). We have proper validations against them at other layers that are reported as compilation errors and once you get to actually using the `Object`, they should never happen.

I think it would be best to take the same approach in yulopti but if there is some specific case where it's problematic, we could raise an exception. Either one of those already available in `Exceptions.h` or a new one defined specifically for this.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +89,70 @@ class YulOpti
+
+		yulAssert(
+			subObject != nullptr,
+			"Assembly object may not contain code."
```
Currently it sounds as if it's not allowed to contain code.
```suggestion
			"Assembly object does not contain code."
```

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +206,138 @@ class YulOpti
 	}
 
-	void disambiguate()
+	void applyFunctionToObject(shared_ptr<Object> _object, function<void(Object&)> _function)
```
```suggestion
	void applyFunctionToObjectAndSubobjects(shared_ptr<Object> _object, function<void(Object&)> _function)
```

cameel replied:
```suggestion
	void applyFunctionToObject(Object& _object, function<void(Object&)> _function)
```

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +206,138 @@ class YulOpti
+		);
+	}
+
+	void runSteps(string _source, string _objectPath, string _steps)
```
```suggestion
	void runSteps(string const& _source, string const& _objectPath, string const& _steps)
```

cameel replied:
Now that you added `runSequence()`, both sound like they would do the same thing. Let's make the difference between them clear:
```suggestion
	void parseAndRunSteps(string _source, string _objectPath, string _steps)
```

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -331,20 +485,25 @@ int main(int argc, char** argv)
+
 		if (!nonInteractive)
-			cout << input << endl;
+			yulOpti.parseAndPrint(input, objectPath);
```
This helper is used only once and I think it's actually clearer to have it directly here since it's so simple:
```suggestion
			{
				yulOpti.parse(input, objectPath);
				yulOpti.printObject();
			}
```

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -331,20 +485,25 @@ int main(int argc, char** argv)
 			return 1;
 		}
 
+		if (arguments.count("object"))
+			objectPath = arguments["object"].as<string>();
```
We should report an error when this is non-empty and the input is a block, not an object.

dflupu replied:
I've added this check in `parse()`

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -284,6 +431,11 @@ int main(int argc, char** argv)
 			)
+			(
+				"object",
+				po::value<string>(),
```
```suggestion
				po::value<string>()->value_name("path"),
```

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -284,6 +431,11 @@ int main(int argc, char** argv)
 			)
+			(
+				"object",
+				po::value<string>(),
+				"path to a yul object in the input"
```
```suggestion
				po::value<string>()->value_name("path"),
				"Dotted path to a specific Yul object in the source. "
				"If not specified, the top-level object is used. "
				"Only valid if the source actually contains objects (rather than a block of Yul code)."
```

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -265,10 +411,11 @@ int main(int argc, char** argv)
 	interactively read from stdin.
 	In non-interactive mode a list of steps has to be provided.
 	If <file> is -, yul code is read from stdin and run non-interactively.
+	An <object> flag may be provided, specifying a dotted path to an object in the input.
```
```suggestion
	An --object flag may be provided, specifying a dotted path to an object in the input.
```
But I think this is not necessary here. The option description below is already enough.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +206,138 @@ class YulOpti
+		if (!m_inputWasCodeBlock)
+			cout << m_object->toString(&m_dialect) << endl;
+		else
+			cout << AsmPrinter{m_dialect}(*m_object->code)  << endl;
```
```suggestion
			cout << AsmPrinter{m_dialect}(*m_object->code) << endl;
```

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -214,47 +359,48 @@ class YulOpti
 private:
-	shared_ptr<yul::Block> m_ast;
+	shared_ptr<yul::Object> m_object;
+	bool m_inputWasCodeBlock = false;
```
Isn't it still a code block? A block should not be converted to an object by the tool.
```suggestion
	bool m_inputIsCodeBlock = false;
```

dflupu replied:
Renamed.

The tool now uses `ObjectParser::parse` which has a special case for code blocks. That makes it possible to handle objects and code blocks the same for the most part. Are there issues with that approach?

cameel replied:
It's fine as long as you don't print an object when the input was a block.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +206,138 @@ class YulOpti
+	void printObject()
+	{
+		if (!m_inputWasCodeBlock)
+			cout << m_object->toString(&m_dialect) << endl;
```
Please assert that the object has no subobjects or data objects in this case.

EDIT: I actually meant the other case, when it's a block. I see I selected the wrong branch.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +89,70 @@ class YulOpti
+				throw runtime_error("Could not parse source.");
 			}
+
+			m_object = getSubObject(object, _objectPath);
```
Don't we have to keep the whole object? If you do this, I'd expect `parse()` to fail if called multiple times (unless the nested object happens to have the same path inside it - but even then it would be a different subobject than requested).

cameel replied:
Ah, I see we don't. We only run `parse()` once. So I guess this is fine as long as a nested object cannot refer to any objects on upper levels and I think that currently it can't.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +206,138 @@ class YulOpti
+		);
+	}
+
+	void runSequence(string_view _steps)
+	{
+		applyFunctionToObject(
+			m_object,
+			[&](Object& _object)
+			{
+				OptimiserSuite{*m_context}.runSequence(_steps, *_object.code);
+			}
+		);
+	}
```
I recently realized that there's a big problem here. This code ignores the distinction between *creation* and *deployed* objects (see [Specification of Yul Object](https://docs.soliditylang.org/en/latest/yul.html#specification-of-yul-object)). Normally the optimizer considers the top-level object and any objects with names not ending with `_deployed` to be creation objects. For these `expectedExecutionsPerDeployment` (i.e. `--optimize-runs`) is set to 1 and it affects the result of the optimization. This logic sits in `YulStack::optimize()` but the code here ignores it and just walks the object treating every sub-object the same way.

This wasn't a problem as long as we were working with only one object in isolation. With this PR it becomes misleading because when you pass it the whole object you'll get slightly different output than you would get from the compiler.

dflupu replied:
I see. Would it make sense to call `YulStack::optimize()` here instead or should I implement the same logic?

I'm assuming that the top-level object will be the top-level object from the input and never the object given by `--object`?

cameel replied:
I'd rather avoid reimplementing all of it because we'll easily forget to update yulopti if the code diverges. On the other hand we already make assumptions about what the optimizer does so I don't think we can fully avoid that.

The part I'd like to avoid duplicating the most is the logic for determining if an object is a creation object. I think it would be acceptable to keep the current solution of walking the object on your own if you extract a common method with logic for determining it and make both yulopti and `YulStack` use it. You could put it in `Object`.

Also, I'd recommend doing this (and the refactor I suggested in https://github.com/ethereum/solidity/pull/12941#discussion_r1018349654) in a separate PR. Changes outside of yulopti (which is just a non-user-facing testing tool) have more potential to break things so would be better to have them separately and then to get more people to look specifically at them.

cameel replied:
> I'm assuming that the top-level object will be the top-level object from the input and never the object given by `--object`?

We could do it either way - with the top being fixed or with it being the subobject one specified by `--object`- but I think the former would be more intuitive. I.e. we always optimize the whole object and `--object` works simply as a filter to show you just the one object you care about - as opposed to cutting out the subobject and optimizing it as if it was the whole thing (which you can always achieve just by modifying the file).

It would be a good idea to say that in the help string of the option.

dflupu replied:
I've updated the PR, adding proper handling of creation objects. For now I've duplicated the check code. I'm not sure how the check can be properly refactored into `Object` since knowing whether an object is a creation object or not requires knowing if the object is at the root of the tree. Some possible method declarations would be:
```
bool Object::isCreation(string _qualifiedPath);
bool Object::isCreation(int depth);
bool Object::isCreation(bool isRootObject);
```

Maybe just `bool Object::isDeployed()` would do?

I had done the refactor suggested in [#12941 (comment)](https://github.com/ethereum/solidity/pull/12941#discussion_r1018349654) on this PR branch previously. I can of course move it into a separate PR.

cameel replied:
> ```solidity
> bool Object::isCreation(bool isRootObject);
> ```

I'd go with this one. The one taking the path might actually be nicer but in some places where we need this (in particular `YulStack`), we don't have the path at hand so we'd have spend extra effort to construct it. We could also add both, one using the other under the hood, but that's optional.

> Maybe just `bool Object::isDeployed()` would do?

If you mean a method only checking for `_deployed` in the name, I think that would be a bit misleading because for the top-level object it could return `true` even if it's a creation object. I see `isDeployed` as the opposite of `isCreation`.

> I had done the refactor suggested in https://github.com/ethereum/solidity/pull/12941#discussion_r1018349654 on this PR branch previously. I can of course move it into a separate PR.

I'd really recommend moving it to `Object` and doing that in a separate PR. We have several places that have this condition hard-coded and I think it would be best if they all used a single method. To be honest, I'm not even confident that they deal with the top-level object the same way so it would be a good opportunity to check that and maybe add some tests to cover it.

[Specification of Yul Object](https://docs.soliditylang.org/en/latest/yul.html#specification-of-yul-object) does not even mention the top-level object so we could also update it in that PR.

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +182,138 @@ class YulOpti
+	{
+		if (!m_inputIsCodeBlock)
+			cout << m_object->toString(&m_dialect) << endl;
+		else {
```
```suggestion
		else
		{
```

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +182,138 @@ class YulOpti
-	void disambiguate()
+	void applyFunctionToObjectAndSubobjects(Object& _object, function<void(Object&)> _function)
+	{
+		for (auto const& subObjectNode: _object.subObjects) {
```
```suggestion
		for (auto const& subObjectNode: _object.subObjects)
		{
```

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -87,37 +88,47 @@ class YulOpti
+
+			if (m_object == nullptr) {
+				throw runtime_error("Assembly object not found.");
 			}
```
```suggestion
			if (m_object == nullptr)
				throw runtime_error("Assembly object not found.");
```
Also, maybe we should throw something a different type such as `util::Exception` or `yul::YulException`  instead of `runtime_error`.

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +182,138 @@ class YulOpti
+			[&](Object& _object)
+			{
+				OptimiserSuite{*m_context}.runSequence(_steps, *_object.code);
+			}
```
```suggestion
			[&](Object& _object) { OptimiserSuite{*m_context}.runSequence(_steps, *_object.code); }
```

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +182,138 @@ class YulOpti
+			[&](Object& _object)
+			{
+				VarNameCleaner::run(*m_context, *_object.code);
+			}
```
```suggestion
			[&](Object& _object) { VarNameCleaner::run(*m_context, *_object.code); }
```

matheusaaguiar commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +182,138 @@ class YulOpti
+			[&](Object& _object)
+			{
+				StackCompressor::run(m_dialect, _object, true, 16);
+			}
```
```suggestion
			[&](Object& _object) { StackCompressor::run(m_dialect, _object, true, 16); }
```

matheusaaguiar reviewed (commented):
Mostly nitpicks about code style.

cameel commented on libyul/Object.h:
```diff
@@ -125,6 +126,13 @@ struct Object: public ObjectNode
+	/// @returns a shared_ptr to the Object or a nullptr if it was not found.
+	static std::shared_ptr<Object> objectAt(
+		std::shared_ptr<Object> const& _object,
+		std::string const& _qualifiedPath
```
I'd call it `qualifiedName` just `pathToSubObject()`:
```suggestion
		std::string const& _qualifiedName
```

But you have a point that it's a path. It's just that in this file by path we seem to mean specifically a vector with indexes so I'd stick to that for consistent terminology.

cameel commented on libyul/Object.h:
```diff
@@ -125,6 +126,13 @@ struct Object: public ObjectNode
+
+	/// Recursively searches for an Object at @param _qualifiedPath within @param _object.
+	/// @returns a shared_ptr to the Object or a nullptr if it was not found.
+	static std::shared_ptr<Object> objectAt(
```
Any reason not to make it a member function and call it `subObjectAt()`? You need an `Object` instance to use it anyway.

cameel replied:
Then I'd also move it next to `pathToSubObject()`.

matheusaaguiar replied:
Marked as resolved, but suggested changes were not applied.

r0qs replied:
Now it is resolved :)

cameel commented on libyul/Object.cpp:
```diff
@@ -144,3 +145,28 @@ vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
+	yulAssert(subObject, "Assembly object <" + subObject->name.str() + "> does not contain code.");
+
+	return objectAt(subObject, subObjectPath);
+}
```
I think it would be better to use `pathToSubObject()` in the implementation here. They both traverse the object tree in the same way and just return different things. Having this not construct a vector of indexes is a bit more efficient but in the grand scheme of things the difference is negligible and being sure that they won't get out of sync is more than worth it.

For example we already have a PR that extends `pathToSubObject()` to support data objects (#13707) and if your function was calling it, we'd get that functionality for free here, without having to do anything.

r0qs replied:
I proposed a solution for this, please see https://github.com/ethereum/solidity/pull/12941/commits/70c816c54bc2a4bfdfa34ae5f42dd7d6ef84a897 :)

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -91,33 +92,32 @@ class YulOpti
+			auto scanner = make_shared<Scanner>(charStream);
+
+			if (!m_inputIsCodeBlock && scanner->currentToken() == Token::LBrace)
+				m_inputIsCodeBlock = true;
```
Any reason this is not updated when `m_inputIsCodeBlock` is already true?
```suggestion
			m_inputIsCodeBlock = (scanner->currentToken() == Token::LBrace);
```

I don't think it makes much difference in the current implementation (it won't randomly flip from one to the other anyway) and just assigning it every time makes the logic clearer.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -171,28 +171,152 @@ class YulOpti
+					_object.qualifiedDataNames()
+				);
+
+				bool success = analyzer.analyze(*_object.code);
+				yulAssert(success && !_errorReporter.hasErrors(), "Invalid assembly/yul code.");
```
Before this PR an analysis error resulted in the errors being printed with `printErrors()`. I think we should keep that (unless I'm missing something and this behavior is somehow preserved?).

It would also throw `std::runtime_error`, which is semantically different from an assertion, even though currently both are being handled the same way in `catch (...)`. We should really refactor this some day to throw a more specific exception class on validation/runtime errors and handle them differently from failed assertions.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -265,7 +382,7 @@ int main(int argc, char** argv)
 			R"(yulopti, yul optimizer exploration tool.
 	Usage: yulopti [Options] <file>
-	Reads <file> as yul code and applies optimizer steps to it,
+	Reads <file> containing a either a yul object or a yul code block and applies optimizer steps to it,
```
```suggestion
	Reads <file> containing either a yul object or a yul code block and applies optimizer steps to it,
```

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -214,47 +338,40 @@ class YulOpti
-				_source = AsmPrinter{m_dialect}(*m_ast);
+
+				resetNameDispenser();
+				runCodeAnalyzer(errorReporter);
```
Rerunning analysis after every step is probably superfluous. Or at least, `yulopti` was not doing that before. Any reason to introduce it now?

r0qs replied:
I removed it since I didn't see a reason for it there either.

NunoFilipeSantos commented:
Hi @dflupu ! 👋 are you planning to come back soon to this PR, or should we close it for now?

matheusaaguiar commented on libyul/Object.cpp:
```diff
@@ -144,33 +145,62 @@ set<YulString> Object::qualifiedDataNames() const
-	vector<size_t> path;
-	Object const* object = this;
-	for (string const& currentSubObjectName: subObjectPathComponents)
+	for (size_t i = 0; i < subObjectPathComponents.size(); i++)
```
why not iterators?

matheusaaguiar commented on libyul/Object.cpp:
```diff
@@ -144,33 +145,62 @@ set<YulString> Object::qualifiedDataNames() const
-	vector<size_t> path;
-	Object const* object = this;
-	for (string const& currentSubObjectName: subObjectPathComponents)
+	for (size_t i = 0; i < subObjectPathComponents.size(); i++)
```
Why not range-based for?

```suggestion
	for (auto const& currentSubObjectName: subObjectPathComponents)
```

r0qs replied:
No specific reason. Fixed :)

cameel commented on libyul/Object.h:
```diff
@@ -116,6 +117,14 @@ struct Object: public ObjectNode
+		Object const* _object,
+		YulString _qualifiedName,
+		std::function<bool(Object const*)> _visitor
+	);
```
- If something requires a single instance of the class as an argument, in most cases it should just be a member function.
- I'd call it `visitPath()`. The current name is a bit ambiguous - to me at first it sounded as if it would visit the whole tree.
- You don't say what happens when the `_visitor` returns `false`. Or when the object at the path does not exist. Or is a `Data` object. Or in which direction the path is traversed.

cameel commented on libyul/Object.cpp:
```diff
@@ -144,33 +145,61 @@ set<YulString> Object::qualifiedDataNames() const
+	traverseObjectTree(_object.get(), YulString(_qualifiedName), [&](Object const* _obj) -> bool {
+		if (_qualifiedName.empty() || _qualifiedName == _obj->name.str())
+		{
+			object = make_shared<Object>(*_obj);
```
`make_shared()` will create a copy of the object. We have all the objects neatly organized in a tree and we don't want extra copies floating around outside, with references to the ones inside. You should avoid casually copying them. That would only be needed if you were going to modify the tree structure.

If you're going to return a `shared_ptr` to an `Object`, it should be pointing at one of the objects in the tree. But in this case, you don't really need `shared_ptr` at all. Just return a pointer or a reference. Use smart pointers in contexts where you need to express ownership. E.g. when declaring a class member that holds the reference to it long-term. Informational functions like this need not be concerned with ownership.

cameel replied:
Also, you have `_object`, `object` and `_obj` here. That does not help understand what each of them is for. Names should be more specific.

r0qs replied:
Right! Thanks, I changed it to a pointer and renamed the other variables :)

cameel commented on libyul/Object.cpp:
```diff
@@ -144,33 +145,61 @@ set<YulString> Object::qualifiedDataNames() const
+{
+	std::shared_ptr<Object> object = nullptr;
+	traverseObjectTree(_object.get(), YulString(_qualifiedName), [&](Object const* _obj) -> bool {
+		if (_qualifiedName.empty() || _qualifiedName == _obj->name.str())
```
Note that `name` is unqualified. I.e. for qualified path `A.B.C` the name is `C`. I don't think this comparison will work as intended.

Also the check for empty qualified path looks like it should be handled before the traversal. Do we even have to treat it as a special case? It looks like it would work regardless.

r0qs replied:
Indeed!! Fixed ;)

cameel commented on libyul/Object.cpp:
```diff
@@ -144,33 +145,61 @@ set<YulString> Object::qualifiedDataNames() const
+		auto subObjectIt = ranges::find_if(
+			_object->subObjects,
+			[&currentSubObjectName](auto const& _subObject) { return _subObject->name.str() == currentSubObjectName; }
 		);
```
This replaces search in a hash map with linear search through a vector. The old version was both more efficient (for large objects) and more concise.

Since this assumes that the key must be present anyway, we could even make the old version more concise by using `operator[]` (and asserting beforehand with `map::count()`) instead of `map::find()`. Then we would not have to deal with iterators here at all.

The `if` below should also be an assert.

r0qs replied:
Done. I added the `operator::at` instead of `operator[]` since, to keep the same semantics of the later, we would need to add a new element if it does not exists.

r0qs commented:
@cameel I noticed that you were reviewing the work in another branch: https://github.com/ethereum/solidity/commits/yulopti-object-support and decided to apply the ones that still make sense in the current state of the PR! Thanks a lot ;)

The only one that I didn't apply was the change suggested in this commit: https://github.com/ethereum/solidity/commit/f34171f8063d1939fd73b525ac5d76267f28b2ed because I use some `std::string` methods like `find_last_of` and `substr` to determine the name of the target object, so if we change the function to receive a `YulString` instead of `std::string` we would need to make conversions between them inside `subObjectAt` anyway.

nikola-matic commented on libyul/Object.cpp:
```diff
@@ -38,6 +38,13 @@ using namespace solidity::langutil;
+
+Object const* Object::at(YulString _name) const
+{
+	size_t subIndex = this->subIndexByName.at(_name);
```
This can throw.

r0qs replied:
Isn't this the expected behavior of `at` operator?

nikola-matic replied:
Yeah. Is it handled properly if it throws?

r0qs replied:
Yes, or at least, I thought so. But looking at the code again, all exceptions in yulopti are handled in the main try-catch loop as expected, although @cameel mentioned that we could improve on that eventually as well https://github.com/ethereum/solidity/pull/12941#discussion_r1103136596. However, as this code is now also shared by `Object::pathToSubObject` through the common `Object::visitPath` this may indeed throw in [EVMDialect](https://github.com/ethereum/solidity/blob/1b5c6f66930509d45c7b97ccd803e649c54a8702/libyul/backends/evm/EVMDialect.cpp#L222) when creating builtins. So yeah, good point. I will check again how the `pathToSubObject` used to work before this change to ensure the same behavior.

r0qs replied:
It appears that the function would return an empty path if there is no [subObjectPathComponents](https://github.com/ethereum/solidity/blob/develop/libyul/Object.cpp#L142) and an assertion failure may occur if the  [subObject is not found](https://github.com/ethereum/solidity/blob/develop/libyul/Object.cpp#L155), which still the case with the current change. Also, there is an assertion for `object->subIndexByName.count(subObjectName)` right before the invocation of [object->at(subObjectname)](https://github.com/ethereum/solidity/pull/12941/files#diff-85786c576d9ffa5957420e25761d465d531c5e9648f3a1d3f5cb91c40e201c5eR153). Therefore, in this scenario, it should not throw an error in the `Object::at` function, instead, it should trigger the assertion failure in `visitPath`.

r0qs replied:
Also, I just found a bug now. If you don't provide an object with the `--object` command-line flag, it should use the top-level object as default but it doesn't.

r0qs replied:
Fixed. I added an assert in any case. You can test with:
```
# raw Yul as input:
./build/test/tools/yulopti test/libyul/yulOptimizerTests/fullSuite/medium.yul
# Yul objects as input:
./build/test/tools/yulopti test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object "A.B" test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object "A.B.C.D" test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object "B.E" test/libyul/objectCompiler/subObjectAccess.yul
```

r0qs commented on libyul/Object.cpp:
```diff
@@ -129,33 +141,64 @@ std::set<YulString> Object::qualifiedDataNames() const
-	for (std::string const& currentSubObjectName: subObjectPathComponents)
+
+	// Iterate over sub objects ignoring the top-level object in the path.
+	for (std::string const& currentSubObjectName: subObjectPathComponents | ranges::views::drop(1))
```
I'm not completely sure if it is fine to drop the first path element in the context of `Object::pathToSubObject`.

r0qs replied:
Actually it is not, because of the use of `starts_with` and `substr` there. I will fix it.

r0qs replied:
It should be fixed now:)

nikola-matic commented on libyul/Object.cpp:
```diff
@@ -31,13 +31,25 @@
+		this->subIndexByName.count(_name),
+		"Assembly object <" + _name.str() + "> not found."
+	);
+	size_t subIndex = this->subIndexByName.at(_name);
+	return dynamic_cast<Object const*>(this->subObjects[subIndex].get());
```
```suggestion
	size_t subIndex = subIndexByName[_name];
	return dynamic_cast<Object const*>(subObjects[subIndex].get());
```

No need for `at` now that you've asserted that the object is indeed present and produces a valid index.

nikola-matic replied:
Actually - keep the `this->`, but none of the members in `Object` are prepended with `m_` to indicate that they're indeed members, which is not great; but this isn't the place to fix it.

r0qs replied:
> No need for `at` now that you've asserted that the object is indeed present and produces a valid index.

Indeed. However, this would make the method non-const. I would prefer to keep it const, but I can make the change if necessary.

> Actually - keep the `this->`, but none of the members in `Object` are prepended with `m_` to indicate that they're indeed members, which is not great; but this isn't the place to fix it.

Yeah, I thought about fix it here, but it is better to open a new PR for that.

nikola-matic replied:
> Indeed. However, this would make the method non-const. I would prefer to keep it const, but I can make the change if necessary.

Ah, you're right, didn't even think about that.

cameel replied:
I'd actually remove `this->`. We never use it explicitly and it was always pointed out in my PRs.

cameel replied:
Also, I don't think `at()` is a good name if this method ignores `Data` objects. It's easy to take it for a general accessor like `map::at()` and assume that the index is empty when `nullptr` is returned.

The fact that this happens is not even documented in its docstring.

nikola-matic commented on libyul/Object.cpp:
```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
+		return true;
+	});
+
+	return foundObject;
```
I find this somewhat counter intuitive, perhaps due to the naming, or due to lack of asserts; in any case, we're making sure that `_qualifiedName` is not empty, so regardless of whether we're using it as is, or as a last qualified name delimited by `.`, we're always going to have a qualified name.

Then we search for the sub-object using that name, and store into `foundObject`, but the same `foundObject` can end up being a `nullptr`. The question now is - do we always expect `_qualifiedName` to produce a valid subjobject - if yes, this should have an assert.

r0qs replied:
Note that `_qualifiedName` is not modified, and this function expects it to be in the format: `(A.*)+`, where `A` is a Yul object name. The search is performed using `_qualifiedName`, and `targetObjectName` is the name of the sub-object we are looking for. It will be `nullptr` only if the object is not found, i.e., it is not a sub-object of the top-level object.

But I don't think we always expect that the `_qualifiedName` will always contain valid sub-objects, since this is an user input. Some asserts are already in place to validate parts of it (see `Object::visitPath`), but without test coverage, I'm unsure if we are covering everything. For instance, these inputs are invalid and fail as expected:

```
./build/test/tools/yulopti --object "A." test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object ".A" test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object "." test/libyul/objectCompiler/subObjectAccess.yul
./build/test/tools/yulopti --object "..." test/libyul/objectCompiler/subObjectAccess.yul
# A is not a sub-object of B
./build/test/tools/yulopti --object "B.A" test/libyul/objectCompiler/subObjectAccess.yul
```

r0qs replied:
Should I validate it with a regex? Although it could be any valid Yul Object identifier, so not really sure if we need more validation than we already have.

nikola-matic replied:
> Should I validate it with a regex? Although it could be any valid Yul Object identifier, so not really sure if we need more validation than we already have.

Nope, no need.

cameel replied:
Having asserts instead of validations here is fine for now - in the compiler we have those validations already elsewhere and yulopti is an internal tool. I mean, I'd include a validation if I were writing it myself, but not having one is not a big deal as long as it has an assert with a good message.

Just, if we're using asserts, I'd try to make it somewhat consistent with how other functions behave. Now `subObjectAt()`, `visitPath()`, and `pathToSubObject()` each has a different idea of what to do on objects that do not exist and on `Data` objects. Since `pathToSubObject()` already existed and we use it elsewhere in the compiler, IMO it would be sensible if all of them followed its behavior (i.e. just assert against all of these cases).

nikola-matic commented:
@cameel there's still a CR from your side on this PR (from Nov 9th).

cameel commented:
I'll give it a final look tomorrow.

cameel commented on libyul/Object.cpp:
```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
-		yulAssert(object->subId != std::numeric_limits<size_t>::max(), "");
-		path.push_back({object->subId});
+
+		YulString subObjectName = YulString{currentSubObjectName};
+		object = object->at(subObjectName);
```
Just a general remark about naming: please avoid renaming things in a way that does not make it clear what the distinction between them really is. Here both `subObjectName` and `currentSubObjectName` are pretty much the same thing and the difference is not that one is not "current". It's that one is a `YulString`. So something like `currentSubObjectNameYul` would be a much clearer name for the former. But in this situation I find that `YulString{currentSubObjectName}` is not much longer so I'd just use it directly:

```suggestion
		object = object->at(YulString{currentSubObjectName});
```

cameel commented on libyul/Object.h:
```diff
@@ -116,6 +117,21 @@ struct Object: public ObjectNode
+
+	/// Visits all subobjects in the path given by the @a _qualifiedName
+	/// of the current object applying the function @a _visitor.
+	/// If @a _visitor returns `true` the visiting stops, otherwise,
+	/// it continues until the last subobject in the path is reached.
```
I'd make it return `false` to stop. That's how it works in AST visitors so doing the opposite here will only make mistakes more likely.

cameel replied:
Also, the docstring does not specify what happens when the path leads to an object that does not exist or is a `Data` object.

cameel commented on libyul/Object.cpp:
```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
+
+		YulString subObjectName = YulString{currentSubObjectName};
+		object = object->at(subObjectName);
+		if (object && _visitor(object))
+			break;
```
You continue the loop when `object` is null and the next cycle may try to access it. This will segfault.

`test.yul`
```yul
object "A" {
    code {}
    data "B" ""
}
```
```
yulopti --input-file test.yul --object A.B.C
```
```
Segmentation fault (core dumped)
```

cameel commented on libyul/Object.cpp:
```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
-std::vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
+void Object::visitPath(YulString _qualifiedName, std::function<bool(Object const*)> const& _visitor) const
 {
-	yulAssert(_qualifiedName != name, "");
-	yulAssert(subIndexByName.count(name) == 0, "");
```
These asserts are still relevant here and should not be moved.

The second one checks an assumption for the `if` below that strips `name` from `_qualifiedName`. The assumption is that an object never contains a sub-object of the same name. If that was not true, such paths would be ambiguous.

The first one ensures that we'll visit at least one subobject, just like `yulAssert(!_qualifiedName.empty(), "");` below.

cameel commented on libyul/Object.cpp:
```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
+
+	Object const* foundObject = nullptr;
+	this->visitPath(_qualifiedName, [&](Object const* _subObject) -> bool {
+		if (targetObjectName != _subObject->name)
+			return false;
```
I don't think this is correct. What if `targetObjectName` appears earlier on the path? E.g. for `A.B.A.B` the function would give you the object corresponding to `A.B`.

You should just let `visitPath()` iterate the whole `_qualifiedName`, just unconditionally updating `foundObject`. When it stops, `foundObject` will be the last object on the path.

cameel commented on libyul/Object.h:
```diff
@@ -116,6 +117,21 @@ struct Object: public ObjectNode
 	std::vector<size_t> pathToSubObject(YulString _qualifiedName) const;
 
+	/// Searches for a subobject at @param _qualifiedName within the current object.
+	/// @returns a pointer to the subobject or a nullptr if it was not found.
```
Please clarify what happens when the path leads to a `Data` object.

Also, it's not entirely true that it returns null if not found. The function has a check against that but it will trigger an assert in `Object::at()` before it even gets to that check.

cameel commented on test/tools/yulopti.cpp:
```diff
@@ -90,33 +91,29 @@ class YulOpti
-				cerr << "Error analyzing source." << endl;
-				printErrors(_charStream, errors);
-				throw std::runtime_error("Could not analyze source.");
+				printErrors(charStream, errors);
+				solThrow(Exception, "Could not parse source.");
```
The exception goes from here to the `catch` block, which will print the errors again. We should print them only once.

Also, `runCodeAnalyzer()` should use `solThrow()` too instead of asserting that it expects no errors (which is not true).

cameel reviewed (changes requested):
There are still significant problems in the changes to `Object`. I found a case where it will segfault and also the new functions are pretty inconsistent in how they behave in corner cases, which may easily lead to future bugs and is a regression compared to the current situation.

I haven't reviewed the yulopti part very thoroughly.

cameel commented on libyul/Object.cpp:
```diff
@@ -129,33 +141,60 @@ std::set<YulString> Object::qualifiedDataNames() const
+	yulAssert(_qualifiedName != name, "");
+	yulAssert(subIndexByName.count(name) == 0, "");
+
+	this->visitPath(_qualifiedName, [&](Object const* _object) -> bool {
+		yulAssert(_object->subId != std::numeric_limits<size_t>::max(), "");
+		path.push_back(_object->subId);
+		return false;
+	});
```
By the way `visitPath()` would be more idiomatic if it used iterators instead of callbacks. It would let us use ranges:

```
visitPath(_qualifiedName) | transform(objectToSubId) | to<vector>();
```

This would make most of these new helpers obsolete. I'd not do it in this PR though. We want to be finally done with it :)
